### ISSUE TITLE
training on mars data set is not completing. HELP

### ISSUE THREAD

**Issue by nirojshrestha019** [nwojke/cosine_metric_learning#87]
Actually, I want to train deep sort on my custom dataset.

But at first, just wanted to make sure that the input data format are correct and codes run perfectly for MARS data set. I have downloaded the MARS dataset and have stored it into **MARS-evaluation-master** folder. This folder consists of **bbox_train** and **bbox_test** folder. I have only used some of the data from the MARs dataset for testing purposes. The total number of training images are:

**Train set size: 3318 images, 12 identities**

I am running the following line of command to train it:

`python train_mars.py \
    --dataset_dir=./MARS-evaluation-master \
    --loss_mode=cosine-softmax \
    --log_dir=./output/mars/ \
    --run_id=cosine-softmax`

**_Could anyone help me to figure out why training is not completed?_**

But the training is not being completed. The code enters into **train_loop** function in **train_app.py** file. There is no issue, the code is running and is stuck in the following line:

Train set size: 3318 images, 12 identities
creating network factory
creating train_loop
read_from_file:  True
WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/train_app.py:245: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/train_app.py:252: The name tf.read_file is deprecated. Please use tf.io.read_file instead.

WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/train_app.py:254: The name tf.image.resize_images is deprecated. Please use tf.image.resize instead.

WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/queued_trainer.py:316: The name tf.FIFOQueue is deprecated. Please use tf.queue.FIFOQueue instead.

WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/train_app.py:268: The name tf.summary.image is deprecated. Please use tf.compat.v1.summary.image instead.

WARNING:tensorflow:From /tensorflow-1.15.2/python3.6/tensorflow_core/contrib/layers/python/layers/layers.py:1057: Layer.apply (from tensorflow.python.keras.engine.base_layer) is deprecated and will be removed in a future version.
Instructions for updating:
Please use `layer.__call__` method instead.
WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/nets/deep_sort/network_definition.py:19: The name tf.get_variable_scope is deprecated. Please use tf.compat.v1.get_variable_scope instead.

WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/nets/deep_sort/network_definition.py:28: The name tf.summary.histogram is deprecated. Please use tf.compat.v1.summary.histogram instead.

feature dimensionality:  128
WARNING:tensorflow:From /tensorflow-1.15.2/python3.6/tensorflow_core/contrib/layers/python/layers/layers.py:1634: flatten (from tensorflow.python.layers.core) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.flatten instead.
WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/nets/deep_sort/network_definition.py:84: calling l2_normalize (from tensorflow.python.ops.nn_impl) with dim is deprecated and will be removed in a future version.
Instructions for updating:
dim is deprecated, use axis instead
WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/nets/deep_sort/network_definition.py:94: calling Constant.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/nets/deep_sort/network_definition.py:97: The name tf.summary.scalar is deprecated. Please use tf.compat.v1.summary.scalar instead.

WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/train_app.py:619: sparse_softmax_cross_entropy (from tensorflow.contrib.losses.python.losses.loss_ops) is deprecated and will be removed after 2016-12-30.
Instructions for updating:
Use tf.losses.sparse_softmax_cross_entropy instead. Note that the order of the logits and labels arguments has been changed.
WARNING:tensorflow:From /tensorflow-1.15.2/python3.6/tensorflow_core/contrib/losses/python/losses/loss_ops.py:409: compute_weighted_loss (from tensorflow.contrib.losses.python.losses.loss_ops) is deprecated and will be removed after 2016-12-30.
Instructions for updating:
Use tf.losses.compute_weighted_loss instead.
WARNING:tensorflow:From /tensorflow-1.15.2/python3.6/tensorflow_core/contrib/losses/python/losses/loss_ops.py:152: div (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Deprecated in favor of operator or tf.math.divide.
WARNING:tensorflow:From /tensorflow-1.15.2/python3.6/tensorflow_core/contrib/losses/python/losses/loss_ops.py:154: where (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
WARNING:tensorflow:From /tensorflow-1.15.2/python3.6/tensorflow_core/contrib/losses/python/losses/loss_ops.py:121: add_loss (from tensorflow.contrib.losses.python.losses.loss_ops) is deprecated and will be removed after 2016-12-30.
Instructions for updating:
Use tf.losses.add_loss instead.
WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/losses.py:142: The name tf.log is deprecated. Please use tf.math.log instead.

WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/train_app.py:274: The name tf.trainable_variables is deprecated. Please use tf.compat.v1.trainable_variables instead.

WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/train_app.py:282: The name tf.train.get_or_create_global_step is deprecated. Please use tf.compat.v1.train.get_or_create_global_step instead.

WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/train_app.py:284: The name tf.losses.get_total_loss is deprecated. Please use tf.compat.v1.losses.get_total_loss instead.

WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/train_app.py:286: The name tf.train.AdamOptimizer is deprecated. Please use tf.compat.v1.train.AdamOptimizer instead.

WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/train_app.py:292: The name tf.losses.get_regularization_loss is deprecated. Please use tf.compat.v1.losses.get_regularization_loss instead.

Run training.......................................
---------------------------------------
Run ID:  cosine-softmax
Log directory:  ./output/mars/cosine-softmax
---------------------------------------
WARNING:tensorflow:From /content/drive/My Drive/cosine_metric_learning/queued_trainer.py:406: The name tf.train.Saver is deprecated. Please use tf.compat.v1.train.Saver instead.

inside run..... try
WARNING:tensorflow:From /tensorflow-1.15.2/python3.6/tensorflow_core/contrib/slim/python/slim/learning.py:742: Supervisor.__init__ (from tensorflow.python.training.supervisor) is deprecated and will be removed in a future version.
Instructions for updating:
Please switch to tf.train.MonitoredTrainingSession
2021-02-21 16:27:35.247596: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2021-02-21 16:27:35.252932: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:983] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-02-21 16:27:35.253343: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1639] Found device 0 with properties: 
name: Tesla P4 major: 6 minor: 1 memoryClockRate(GHz): 1.1135
pciBusID: 0000:00:04.0
2021-02-21 16:27:35.253595: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
2021-02-21 16:27:35.254914: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2021-02-21 16:27:35.256392: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcufft.so.10
2021-02-21 16:27:35.256695: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcurand.so.10
2021-02-21 16:27:35.258443: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusolver.so.10
2021-02-21 16:27:35.259321: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusparse.so.10
2021-02-21 16:27:35.262887: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2021-02-21 16:27:35.263010: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:983] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-02-21 16:27:35.263434: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:983] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-02-21 16:27:35.263768: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1767] Adding visible gpu devices: 0
2021-02-21 16:27:35.268516: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2299995000 Hz
2021-02-21 16:27:35.268722: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x2650f40 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2021-02-21 16:27:35.268749: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2021-02-21 16:27:35.347348: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:983] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-02-21 16:27:35.347900: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x2650d80 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2021-02-21 16:27:35.347933: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Tesla P4, Compute Capability 6.1
2021-02-21 16:27:35.348110: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:983] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-02-21 16:27:35.348587: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1639] Found device 0 with properties: 
name: Tesla P4 major: 6 minor: 1 memoryClockRate(GHz): 1.1135
pciBusID: 0000:00:04.0
2021-02-21 16:27:35.348666: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
2021-02-21 16:27:35.348694: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2021-02-21 16:27:35.348719: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcufft.so.10
2021-02-21 16:27:35.348742: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcurand.so.10
2021-02-21 16:27:35.348762: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusolver.so.10
2021-02-21 16:27:35.348782: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusparse.so.10
2021-02-21 16:27:35.348803: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2021-02-21 16:27:35.348879: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:983] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-02-21 16:27:35.349312: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:983] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-02-21 16:27:35.349688: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1767] Adding visible gpu devices: 0
2021-02-21 16:27:35.349766: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
2021-02-21 16:27:35.350789: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1180] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-02-21 16:27:35.350818: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1186]      0 
2021-02-21 16:27:35.350829: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1199] 0:   N 
2021-02-21 16:27:35.350948: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:983] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-02-21 16:27:35.351369: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:983] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-02-21 16:27:35.351766: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:39] Overriding allow_growth setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
2021-02-21 16:27:35.351813: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1325] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 7031 MB memory) -> physical GPU (device: 0, name: Tesla P4, pci bus id: 0000:00:04.0, compute capability: 6.1)
WARNING:tensorflow:From /tensorflow-1.15.2/python3.6/tensorflow_core/python/training/saver.py:1069: get_checkpoint_mtimes (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file utilities to get mtimes.
2021-02-21 16:27:36.991026: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2021-02-21 16:27:37.340760: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7

**Comment by chaaban17** [comment]
@nirojshrestha019 Did u try to monitor your results on tensorboard? The code is not stuck, it just doesn't print anything on terminal.

**Comment by chaaban17** [comment]
> @nirojshrestha019 Did u try to monitor your results on tensorboard? The code is not stuck, it just doesn't print anything on terminal.

also, the code runs indefinitely if you dont set the number of steps, I found this inside the train_app.py:  "--number_of_steps", help="Number of train/eval steps. If None given, "
        "runs indefenitely", default=None, type=int)

Hope this helps.

**Comment by nirojshrestha019** [comment]
Hi @chaaban17 ,

Thank you so much for the response. You are right, the training was running for infinite times. I did change inside the train_app.py: "--number_of_steps", help="Number of train/eval steps", **default=200**, type=int)  as you mentioned and the training completed.

I did try to check tensorboard in colab but was not able to see any graphs. I guess TensorFlow 1.x doesn't support tensorboad well in colab. I did find the solution which works perfectly to display tensorboard in colab:
```!wget https://bin.equinox.io/c/4VmDzA7iaHb/ngrok-stable-linux-amd64.zip
!unzip ngrok-stable-linux-amd64.zip


get_ipython().system_raw('tensorboard --logdir ./output/mars/cosine-softmax/ --host 0.0.0.0 --port 6006 &')
get_ipython().system_raw('./ngrok http 6006 &')

!curl -s http://localhost:4040/api/tunnels | python3 -c \
 "import sys, json; print(json.load(sys.stdin)['tunnels'][0]['public_url'])"
```

Reference: https://stackoverflow.com/questions/60458024/tensorboard-in-tensorflow-1-using-google-colab

Thank you once again.

Regards,
Niroj Shrestha
